### PR TITLE
Added support for Tapas Model

### PR DESCRIPTION
**Pull request by JuheonChu**
# What does this PR do?

Fixes # 20372 

## Before submitting
- This PR adds new support for BetterTransformer integration for the Tapas model.
- This PR adds documentation that indicates BetterTransofrmer integration for Tapas is added.

## Questions
- Can I ask you how I can test to add Bettertransformer feature for Tapas Model?

To: @younesbelkada, @sgugger

### CODE REVIEW

**Comment by JuheonChu** [comment]
Do you mind if I ask you that I am on the right track?

File: tests/bettertransformer/test_bettertransformer_encoder.py
Comment (michaelbenayoun) [inline tests/bettertransformer/test_bettertransformer_encoder.py:283]:
```suggestion
        input_dict = {
```

File: optimum/bettertransformer/models/encoder_models.py
Comment (michaelbenayoun) [inline optimum/bettertransformer/models/encoder_models.py:80]:
```suggestion
    def forward(self, hidden_states, attention_mask, *_, **__):
```

File: optimum/bettertransformer/models/__init__.py
Comment (michaelbenayoun) [inline optimum/bettertransformer/models/__init__.py:27]:
```suggestion
```

**Review by michaelbenayoun — COMMENTED** [review]
For me it looks like you did everything well!
Left a few nits and comments. 
@younesbelkada will be back on Wednesday, so we should wait for his approval, but great work!

**Comment by JuheonChu** [comment]
Thanks for your suggestion and feedback! I reformatted with black, and can I proceed with the CircleCI test one more time to check if my reformatting file works?

File: tests/bettertransformer/test_bettertransformer_encoder.py
Comment (sgugger) [inline tests/bettertransformer/test_bettertransformer_encoder.py:94]:
```suggestion
            elif layer_class == "TapasLayer":
                class_name = "Tapas"
```
No need for the new lines.

**Review by sgugger — APPROVED** [review]
Thanks for your contribution! @younesbelkada good to merge if you are satisfied.

**Review by fxmarty — APPROVED** [review]
Waiting for the tests to run but it looks good to me!

**Comment by fxmarty** [comment]
Apparently a test is not passing: https://github.com/huggingface/optimum/actions/runs/3566116329/jobs/5992236636

You can run them with `pytest tests/bettertransformer/test_XXX.py` -k "name_of_test" --exitfirst -s` locally to see if a specific test passes.

**Comment by JuheonChu** [comment]
@fxmarty I applied `make style` command and made commits with it! 
Thanks for your suggestion and I will try to test it out based on your suggestion.
I sincerely appreciate with your guidance.

**Comment by JuheonChu** [comment]
It seems to me that module packages that  _ tests/bettertransformer/test_bettertransformer_encoder.py_ are importing all fail. However, I would not like to change the structure of a file system based on the suggestion from [here](https://stackoverflow.com/questions/41748464/pytest-cannot-import-module-while-python-can). May I ask you for any suggestions? 

Thank you for your time.

File: optimum/bettertransformer/models/encoder_models.py
Comment (fxmarty) [inline optimum/bettertransformer/models/encoder_models.py:21]:
```suggestion
    def __init__(self, tapas_layer, config):
```

Should be this :D

File: optimum/bettertransformer/models/encoder_models.py
Comment (fxmarty) [inline optimum/bettertransformer/models/encoder_models.py:20]:
```suggestion
class TapasLayerBetterTransformer(BetterTransformerBaseLayer):
```

We are calling `super()` in the init so we should inherit from `BetterTransformerBaseLayer`! 👍

Comment (JuheonChu) [reply optimum/bettertransformer/models/encoder_models.py]: Thank you for your suggestion! That makes a lot of sense!

File: tests/bettertransformer/test_bettertransformer_encoder.py
Comment (younesbelkada) [inline tests/bettertransformer/test_bettertransformer_encoder.py:52]: Could you move this to `ALL_ENCODER_MODELS_TO_TEST` instead? `Tapas` is an encoder-only model ;)

File: tests/bettertransformer/test_bettertransformer_encoder.py
Comment (younesbelkada) [inline tests/bettertransformer/test_bettertransformer_encoder.py:92]: Indeed this is not needed, you can safely remove it !
```suggestion
```

**Review by younesbelkada — COMMENTED** [review]
Thanks so much for your efforts @JuheonChu !
I left few comments, I think that the conversion is currently failing because the keys that you are trying to retrieve (for eg `tapas_layer.attention.query.weight` does not match, here you should use `tapas_layer.attention.self.query.weight` instead. Check the detailed modeling file [here](https://github.com/huggingface/transformers/blob/28247e78819ab9756b81f8df39611c333d099400/src/transformers/models/tapas/modeling_tapas.py). )
One way to make sure that your conversion worked is to constantly run this snippet and debug further 💪 : 
```

from transformers import AutoModel
from optimum.bettertransformer import BetterTransformer

model_id = "hf-internal-testing/tiny-random-TapasModel"
model = AutoModel.from_pretrained(model_id)
bt_model = BetterTransformer.transform(model)
```
Thanks and let us know if you face into any issue!

File: optimum/bettertransformer/models/encoder_models.py
Comment (fxmarty) [inline optimum/bettertransformer/models/encoder_models.py:34]: looking at the Tapas implementation, https://github.com/huggingface/transformers/blob/28247e78819ab9756b81f8df39611c333d099400/src/transformers/models/tapas/modeling_tapas.py#L442 , I think we need here and in the rest `tapas_layer.attention.self.query.weight`, that is why the test is failing!

**Comment by fxmarty** [comment]
Actually, looking at https://github.com/huggingface/transformers/blob/28247e78819ab9756b81f8df39611c333d099400/src/transformers/models/tapas/modeling_tapas.py#L539

I think the `TapasLayerBetterTransformer` is not needed. I think we simply need to use the mapping `"TapasLayer": BetLayerBetterTransformer,`

What do you think @younesbelkada ?

File: tests/bettertransformer/test_bettertransformer_encoder.py
Comment (fxmarty) [reply tests/bettertransformer/test_bettertransformer_encoder.py:52]: Reading https://huggingface.co/docs/transformers/v4.24.0/en/model_doc/tapas#transformers.TapasModel

> The model can behave as an encoder (with only self-attention) as well as a decoder, in which case a layer of cross-attention is added between the self-attention layers, following the architecture described in [Attention is all you need](https://arxiv.org/abs/1706.03762) by Ashish Vaswani, Noam Shazeer, Niki Parmar, Jakob Uszkoreit, Llion Jones, Aidan N. Gomez, Lukasz Kaiser and Illia Polosukhin.

File: optimum/bettertransformer/models/encoder_models.py
Comment (JuheonChu) [reply optimum/bettertransformer/models/encoder_models.py:34]: Oh........ thanks you for providing me with your insights!

File: tests/bettertransformer/test_bettertransformer_encoder.py
Comment (JuheonChu) [reply tests/bettertransformer/test_bettertransformer_encoder.py]: Oh I see! Thanks for an insightful explanation!  I just moved it! :)

**Comment by younesbelkada** [comment]
yes probably, thanks for the heads up @fxmarty !
@JuheonChu , you can also modify the mapping with `"TapasLayer": BertLayerBetterTransformer` and let us know if the conversion works fine!

**Comment by JuheonChu** [comment]
Just made changes accordingly! @younesbelkada  Thank you so much! I will try the conversions and let you know!

File: optimum/bettertransformer/models/__init__.py
Comment (fxmarty) [inline optimum/bettertransformer/models/__init__.py:75]: to be moved in "Bert Family" above

File: optimum/bettertransformer/models/encoder_models.py
Comment (fxmarty) [inline optimum/bettertransformer/models/encoder_models.py:20]: To remove

**Comment by fxmarty** [comment]
Seems all good now @JuheonChu ! Thank you for this PR! So in the end using `BertLayerBetterTransformer` was enough to work with Tapas. If you can just address the points above, and we are ready to merge :)

**Comment by JuheonChu** [comment]
Thank you so much, everyone!! @fxmarty @younesbelkada @sgugger @michaelbenayoun.
I just made changes accordingly! @fxmarty

File: optimum/bettertransformer/models/__init__.py
Comment (fxmarty) [inline optimum/bettertransformer/models/__init__.py:31]: Can you remove `# Tapas Model` and append TapasLayer to the section just below `# Bert Family`?

File: optimum/bettertransformer/models/encoder_models.py
Comment (fxmarty) [inline optimum/bettertransformer/models/encoder_models.py:20]: I meant to remove the class `TapasLayerBetterTransformer` in full 😅 It is not needed in the end since `BertLayerBetterTransformer` can handle Tapas.

Comment (JuheonChu) [reply optimum/bettertransformer/models/encoder_models.py:20]: I think I removed `class TapasLayerBetterTransformer`. Is there anything to be adjusted in this file?

Comment (fxmarty) [reply optimum/bettertransformer/models/encoder_models.py:20]: On your branch, with optimum dev installed (`pip install -e .` in the optimum top folder), can you try to run the script @younesbelkada provided you and see what happens?

```python
from transformers import AutoModel
from optimum.bettertransformer import BetterTransformer

model_id = "hf-internal-testing/tiny-random-TapasModel"
model = AutoModel.from_pretrained(model_id)
bt_model = BetterTransformer.transform(model)
```

**Comment by younesbelkada** [comment]
Hey @JuheonChu ! 
I have pushed directly on your branch ;) The tests should now pass 💪

**Review by younesbelkada — APPROVED** [review]
Thanks a lot @JuheonChu for adding `Tapas` support for `BetterTransformer`! Great job!
Looking forward to your next contributions 💪

**Review by michaelbenayoun — APPROVED** [review]
LGTM!

**Comment by JuheonChu** [comment]
@younesbelkada May I ask you where I can check the merged BetterTransformer feature of Tapas model?

**Pull request merged**
